### ISSUE TITLE
[Feature] Better reusability of NgDocActions.demo function

### ISSUE THREAD

**Issue by Nxtivision** [ng-doc/ng-doc#77]
### Description

The feature `Demo` is really cool, but in some cases, if I want to change only one property to show something to my end users, I will have to recreate a component and change that property manually.

## Example
Here is a `FlatButtonExampleComponent` where the HTML is:
```html
<div class="gap-between">
      <!-- NgDocHTMLSnippetStart(HTML) -->
      <button mat-flat-button color="primary" [disabled]="disabled">
        Primary
      </button>
      <button mat-flat-button color="accent" [disabled]="disabled">
        Accent
      </button>
      <button mat-flat-button color="warn" [disabled]="disabled">Warn</button>
      <!-- NgDocHTMLSnippetEnd(HTML) -->
</div>
```
It displays the result correctly, but when I have to display the result with `disabled` to `true`, I have no way to make the change.

I could use the Playground, of course, but then I will lose the HTML snippet provided inside the Component (it will only display the selector provided in the property `template`)

### Proposed solution

One way to resolve this would be to create an option inside the function `demo` like this:

```nunjucks
{{ NgDocActions.demo(FlatButtonsComponent, { inputs: { disabled: true } }) }}
```

And this will also adapt the snippet, if possible.

Thanks to this new option, I will be able to reuse my demo through the page

### Alternatives considered

I could also use the snippets inside the component and show how it works with or without the disabled property, which is also a way to resolve my issue

**Comment by skoropadas** [comment]
I like the idea of using `inputs` for the demo, but there is another question. If I understand correctly, you have the following scenario: you have two demos, in one you want to show a button with the disabled state, and in the other you don't. That's why you would like an additional "inputs" property to dynamically determine the value of "disabled" input. However, in this case, I assume your template in the demo will look the same:

```html
<button mat-flat-button color="accent" [disabled]="disabled">
        Accent
</button>
```

Will it be clear to users that the value of the variable "disabled" has been changed? Furthermore, if you also show the class body in the demo, it might confuse users because the default value of "disabled" could be false, while the button is disabled.

**Comment by skoropadas** [comment]
While thinking about possible solutions, I remembered that there is another feature for the demo that I wanted to implement. If you take a look at the example through [the link](https://tinkoff.github.io/maskito/core-concepts/overwrite-mode#shift), you will see that there is a demo with examples for several frameworks. Although the displayed demo is written in Angular, I'm thinking about the possibility of doing something similar for NgDoc, so that unrelated files can be added to the list of examples.

Perhaps this feature could also help in your case. For example, you could display a dummy template in the demo code that is not directly related to your demo component and hide the main code of the demo component. In this way, you could manipulate the demo using the inputs field you suggested.

Here's how it could look:

```markdown
{{ NgDocActions.demo("DemoComponent", {
   files: [{name: "HTML", path: "./fake-template1.html"}],
   inputs: {disabled: true},
} }}
```

**Comment by Nxtivision** [comment]
After reading my first message, I saw I wasn't very clear in my example. Here is the complete example used for the FlatButtonsComponent:

```typescript
import { Component, Input } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';

@Component({
  selector: 'public-doc-demo-flat-buttons',
  template: `
    <div class="gap-between">
      <!-- NgDocHTMLSnippetStart(HTML) -->
      <button mat-flat-button color="primary" [disabled]="disabled">
        Primary
      </button>
      <button mat-flat-button color="accent" [disabled]="disabled">
        Accent
      </button>
      <button mat-flat-button color="warn" [disabled]="disabled">Warn</button>
      <!-- NgDocHTMLSnippetEnd(HTML) -->
    </div>
  `,
  standalone: true,
  imports: [MatButtonModule],
  styles: [
    `
      .gap-between {
        gap: 8px;
        display: flex;
      }
    `,
  ],
})
export class FlatButtonsComponent {
  @Input() disabled = false;
}
```

As you can see, the input `disabled` is provided with a default value, like we do in another example with a Playground.

IMHO, it seems we have two different features:

## Input option
The first one should work the same as the inputs provided inside the playground. I have a good example for that:
![2023-07-05_12h04_32](https://github.com/ng-doc/ng-doc/assets/89700030/d4f64b5f-6c59-40e2-9fc1-155a113682d6)

When no option is set (or use the default one), you managed to not show them in the generated template, which is really clever. 
So, if no input option is provided for this property, maybe we can keep the original template, and once we set it, the template will change accordingly, like this:
![2023-07-05_12h10_05](https://github.com/ng-doc/ng-doc/assets/89700030/1af3a9a1-a1a3-47f7-bdbf-dcdde0733ad9)

## Files option
This new option could be really cool if we want to present several way to implement the same thing, like in your example. It could possibly resolve my issue, but I don't think I really get how `fake-template.html` could show to my end users the different states (with or without the disabled property).
Or we can improve the example like this:

```typescript
export interface NgDocDemoActionFilesOptions {
  /** Title for the tab containing the file or retrieved from snippet */ 
  name?: string;
  /** relative path to the alternative example */
  path: string;
}

export interface NgDocDemoActionOptions {
  /** Display demo in the container (true by default) */
  container?: boolean;
  /** Specifies whether the code preview should be expanded (false by default) */
  expanded?: boolean;
  /** Tab name that should be opened by default */
  defaultTab?: string;
  /** List of tabs that should be displayed if they are not empty and exist */
  tabs?: string | string[];
  /** Class name that should be added to the container */
  class?: string | string[];
  /** Alternatives for the current example */
  files?: NgDocDemoActionFilesOptions[];
}
```

I added `retrieved from snippet` because this feature is really interesting. So, in my `fake-template.html`, I can have something like:
```html
<!-- NgDocHTMLSnippetStart(Normal state) -->
<button mat-flat-button color="primary">My button</button>
<!-- NgDocHTMLSnippetEnd(Normal state) -->
<!-- NgDocHTMLSnippetStart(Disabled state) -->
<button mat-flat-button color="primary" [disabled]="true">My button</button>
<!-- NgDocHTMLSnippetEnd(Disabled state) -->
```

However, it could be interesting to have the corresponding view once we click on the tab. In my opinion, it's way more complicated to use this feature instead of the first one. The example for the second solution is not suitable and could be interesting for something else :)

**Comment by skoropadas** [comment]
@Nxtivision As far as I understand, you would like NgDoc to modify the template code of your demo based on the value of the input. For example, if you have a similar template for the demo:

```html
<div class="gap-between">
  <!-- NgDocHTMLSnippetStart(HTML) -->
  <button mat-flat-button color="primary" [disabled]="disabled">
    Primary
  </button>
  <button mat-flat-button color="accent" [disabled]="disabled">
    Accent
  </button>
  <button mat-flat-button color="warn" [disabled]="disabled">Warn</button>
  <!-- NgDocHTMLSnippetEnd(HTML) -->
</div>
```

Currently, NgDoc will display it like this:

```html
<button mat-flat-button color="primary" [disabled]="disabled">
  Primary
</button>
<button mat-flat-button color="accent" [disabled]="disabled">
  Accent
</button>
<button mat-flat-button color="warn" [disabled]="disabled">Warn</button>
```

However, if you had the ability to add and manipulate inputs, then setting the input `disabled=true` would result in NgDoc displaying the template like this:

```html
<button mat-flat-button color="primary" [disabled]="true">
  Primary
</button>
<button mat-flat-button color="accent" [disabled]="true">
  Accent
</button>
<button mat-flat-button color="warn" [disabled]="true">Warn</button>
```

In other words, it would replace the variable in the template with the current value of the input. Did I understand you correctly?

**Comment by Nxtivision** [comment]
Exactly, and moreover, if inside my demo component, I set the default value for my Input() `disabled`, I would like to not show the property disabled inside the template, same behavior as Playground.
So, without manually setting the disabled property would display the snippet like this:
```html
<button mat-flat-button color="primary">
  Primary
</button>
<button mat-flat-button color="accent">
  Accent
</button>
<button mat-flat-button color="warn">Warn</button>
```

Do you think it could be possible ? Maybe reusing the famous `ng-doc-selector` ?

**Comment by skoropadas** [comment]
I really like this idea, as it could indeed make demos very dynamic and reusable. However, it is quite complex and creates several challenges. For example:

1. Although your examples use snippets, this feature should also support standard usage without them. In other words, if you're not using snippets, NgDoc displays the component code as-is. If it's a single TypeScript file, only that file will be displayed. If the component consists of a template and styles file, NgDoc will load and display them as additional tabs in the demo. Here, I see at least one challenge when it's just a TypeScript file. In that case, I would need to retrieve its template specifically and modify it with some values. Loading the file as text and replacing occurrences through regular expressions is not very reliable and can lead to complex implementation. Therefore, I'll likely need to load and modify it using a TypeScript Transpiler. At the very least, I'll need it to extract the template value and update it without accidentally modifying parts of the TypeScript code.

2. The replacement of input property usages in HTML may also not be very reliable. As you understand, I don't have the Angular Language Service, so all I can do is find the property and replace it using standard HTML or regular expressions. However, this approach may have some limitations, but overall, it should work.

Overall, I would like to think about this idea and try to implement it. 

I also thought that, in general, what you want is essentially a playground, but without the sidebar. It might be possible to add the ability to hide the sidebar and pass input property values using the `NgDocActions.playground()` function. This could provide the result you want, at least in the case of buttons.

However, in some cases, when your demo requires proper component initialization or additional TypeScript code, your suggestion with demos would be more useful because when using playgrounds, you only have access to the template.

**Comment by Nxtivision** [comment]
I can imagine how challenging this feature could be, but don't worry: this is not urgent, it exists some workarounds and so, it is not blocking 😉 
Take your time to find the best solution 😉

**Comment by skoropadas** [comment]
Thank you, but in any case, you have given me some improvements that I have almost finished. Now you will be able to further configure playgrounds both in the `ng-doc.page.ts` page configuration and in the template through the `NgDocActions.playground` function. The possible configuration will look like this:

```typescript
export interface NgDocPlaygroundOptions {
	/**
	 * List of selectors that will be used to render the playground (renders all selectors by default)
	 */
	selectors?: string | string[];
	/**
	 * Specifies whether the side panel should be hidden (false by default)
	 */
	hideSidePanel?: boolean;
	/**
	 * Specifies input values for the playground that will be set to the playground component.
	 * These values will be used only once, when the playground is rendered.
	 * If the user resets the playground, these values will be overridden by default values.
	 */
	inputs?: Record<string, unknown>;
	/**
	 * Specifies default values for the playground inputs that will be used for the side panel instead of the default
	 * values from the component. These values **don't override** the default values of the playground component.
	 * They will be used only for initialization and the "Reset" button in the side panel.
	 */
	defaults?: Record<string, unknown>;
}
```

**Comment by skoropadas** [comment]
I'm not sure about adding the `files` property for demos at the moment because the configuration you provide in the `NgDocActions.demo` method relates to how demos are displayed in the application, and none of the properties affect the builder's functionality or how demos are built.

However, I have added the ability to load files for demos using snippets, I think this will meet your needs. In any case, if this is insufficient or not entirely convenient, we can consider adding the `files` property in the future.

**Comment by Nxtivision** [comment]
I will test this new ability, I'll let you know ;) We can probably close this issue, and if I see something wrong, we can reopen it.

You're doing a very great job, we are all proud of you! And my team is absolutely glad with this tool, after the terrible experience in Storybook... 😄

**Comment by skoropadas** [comment]
@Nxtivision Thank you very much! 🥰 There are still a lot of things that I would like to do here, including rewriting the builder and covering it with tests, as it's not very stable at the moment.

Regarding the task - I actually considered this ticket to be more about substituting values in the HTML, which I plan to do in the future, so I didn't close it. For now, I've only added the ability to load files using the snippet, and that's as far as I understand, just a part of this task :)

**Comment by R-Bower** [comment]
@skoropadas I have a reusable component that I use on most of my pages for property documentation.  I pass it a name which is resolved by an external service for props generation.  Right now, I have to create a single component per instance and register that component with the current ng-doc page, then I call `NgDocActions.demo` as normal.  It'd be great if I could re-use the same component and pass it a `name` property from the `NgDocActions.demo` statement in the markdown.  Do you think this will be feasible anytime soon?

**Comment by skoropadas** [comment]
@R-Bower could you provide some examples? It's just a little bit hard to understand what you're trying to achieve

**Comment by R-Bower** [comment]
@skoropadas 

Basically this, but for demos instead.

// currently available for playgrounds:
```
{{ NgDocActions.playground("ButtonPlayground", {hideSidePanel: true, selectors: "button[ng-doc-button-flat]", inputs: {size: "small"}, data: {label: "Small Button"} }) }}
```

// desired for demos
```
{{ NgDocActions.demo("ButtonDemo", {inputs: {size: "small"}} }) }}
```
